Mathlib/Analysis/Calculus/Gradient/Basic.lean

```diff
@@ -0,0 +1,321 @@
+  exact h.hasFDerivAt
+
+theorem HasGradientAt.differentiableAt (h : HasGradientAt f f' x) :
+    DifferentiableAt 𝕜 f x := by use ((toDual 𝕜 F) f'); apply HasGradientAt_iff_HasFDerivAt.mp h
+
+theorem HasGradientWithinAt.differentiableWithinAt (h : HasGradientWithinAt f f' s x) :
+    DifferentiableWithinAt 𝕜 f s x := HasFDerivWithinAt.differentiableWithinAt h
+
+@[simp]
+theorem hasGradientWithinAt_univ : HasGradientWithinAt f f' univ x ↔ HasGradientAt f f' x := by
+  rw [HasGradientWithinAt_iff_HasFDerivWithinAt, HasGradientAt_iff_HasFDerivAt]
+  rw [hasFDerivWithinAt_univ]
+
+theorem DifferentiableOn.hasGradientAt (h : DifferentiableOn 𝕜 f s) (hs : s ∈ 𝓝 x) :
+    HasGradientAt f (∇ f x) x := (h.hasFDerivAt hs).hasGradientAt
+
+theorem HasGradientAt.gradient (h : HasGradientAt f f' x) : ∇ f x = f' :=
+  h.differentiableAt.hasGradientAt.unique h
+
+theorem gradient_eq {f' : F → F} (h : ∀ x, HasGradientAt f (f' x) x) : ∇ f = f' :=
+  funext fun x => (h x).gradient
+
+variable {g : 𝕜 → 𝕜} {g' u : 𝕜} {L' : Filter 𝕜}
+
+theorem toDual_eq_StarRingEnd : ((toDual 𝕜 𝕜) g') 1 = starRingEnd 𝕜 g' := by simp
+
+theorem StarRingEnd_eq_toDual : ((toDual 𝕜 𝕜) (starRingEnd 𝕜 g')) 1 = g' := by simp
+
+theorem Mul_one_eq_SterRingEnd (g' : 𝕜) : ContinuousLinearMap.smulRight (1 : 𝕜 →L[𝕜] 𝕜)
+    (starRingEnd 𝕜 g') = (toDual 𝕜 𝕜) g' := by
+  refine Iff.mpr ContinuousLinearMap.ext_iff ?_
+  simp; intro v; rw [toDual_apply, IsROrC.inner_apply, mul_comm]
+
+theorem SterRingEnd_eq_Mul_one (g' : 𝕜) : ContinuousLinearMap.smulRight (1 : 𝕜 →L[𝕜] 𝕜)
+    g' = (toDual 𝕜 𝕜) (starRingEnd 𝕜 g') := by
+  refine Iff.mpr ContinuousLinearMap.ext_iff ?_
+  simp; intro; rw [toDual_apply, IsROrC.inner_apply, mul_comm]
+  rw [RingHomCompTriple.comp_apply, RingHom.id_apply]
+
+theorem HasGradientAtFilter.hasDerivAtFilter (h : HasGradientAtFilter g g' u L') :
+    HasDerivAtFilter g (starRingEnd 𝕜 g') u L' := by
+  rw [HasDerivAtFilter, Mul_one_eq_SterRingEnd]; exact h
+
+theorem HasDerivAtFilter.hasGradientAtFilter (h : HasDerivAtFilter g g' u L') :
+    HasGradientAtFilter g (starRingEnd 𝕜 g') u L' := by
+  rw [HasGradientAtFilter, ← SterRingEnd_eq_Mul_one]; exact h
+
+theorem HasGradientAt.hasDerivAt (h : HasGradientAt g g' u) :
+    HasDerivAt g (starRingEnd 𝕜 g') u := by
+  rw [HasGradientAt_iff_HasFDerivAt, hasFDerivAt_iff_hasDerivAt, toDual_eq_StarRingEnd] at h
+  exact h
+
+theorem HasGradientAt.hasDerivAt' {g : ℝ → ℝ} {g' u: ℝ} (h :HasGradientAt g g' u) :
+    HasDerivAt g g' u := h.hasDerivAt
+
+theorem HasDerivAt.hasGradientAt (h : HasDerivAt g g' u) :
+    HasGradientAt g (starRingEnd 𝕜 g') u := by
+  rw [HasGradientAt_iff_HasFDerivAt, hasFDerivAt_iff_hasDerivAt, StarRingEnd_eq_toDual]
+  exact h
+
+theorem HasDerivAt.hasGradientAt' {g : ℝ → ℝ} {g' u: ℝ} (h :HasDerivAt g g' u) :
+    HasGradientAt g g' u := h.hasGradientAt
+
+theorem gradient_deriv : ∇ g u = starRingEnd 𝕜 (deriv g u) := by
+  by_cases h: DifferentiableAt 𝕜 g u
```

Comment:
You shouldn't need to distinguish between the two cases, as `gradient` has been defined precisely so that the equality holds.

Reply:
Thank you for your advice. I think we have the need to distinguish whether it is differentiable or not here. The same as the defition of deriv or fderiv, the definition of gradient does not contain the differentiablity. For the convience of prove, I think it is intuitive to think consider the cases.